### ISSUE TITLE
welfordreduce slows down forward layernorm in a bunch of cases

### ISSUE THREAD

**Issue by Chillee** [pytorch/pytorch#120184]
### 🐛 Describe the bug

Compiled (welford) is the default execution, the other plot is with welford reductions turned off.

![image](https://github.com/pytorch/pytorch/assets/6355099/20f6f3c9-0763-4a6e-9b5c-4121d99bd523)

```
import torch
from torch import nn
import torch.nn.functional as F

def bench(f, name=None, iters=1000, warmup=5, display=True, profile=False):
    import time
    from triton.testing import do_bench

    for _ in range(warmup):
        f()
    if profile:
        with torch.profiler.profile() as prof:
            f()
        prof.export_chrome_trace(f"{name if name is not None else 'trace'}.json")

    us_per_iter = do_bench(lambda: f())*1000

    if name is None:
        res = us_per_iter
    else:
        res= f"{name}: {us_per_iter:.3f}us"

    if display:
        print(res)
    return res

class Layer(nn.Module):
    def __init__(self):
        super().__init__()
        dim = 2048
        self.mods = nn.Sequential(nn.Linear(dim, dim), nn.LayerNorm([dim]))

    def forward(self, x):
        return self.mods(x)

class Layered(nn.Module):
    def __init__(self):
        super().__init__()
        self.layers = nn.Sequential(*[Layer() for _ in range(1)])

    def forward(self, x):
        return self.layers(x)

for D in range(256, 8192, 256):
    inp = torch.randn(2048*128, D, dtype=torch.bfloat16).cuda()
    mod = nn.LayerNorm([D], dtype=torch.bfloat16).cuda()

    comp_mod = torch.compile(mod, dynamic=False)
    print(D)
    bench(lambda: mod(inp))
    bench(lambda: comp_mod(inp))
```

cc: @peterbell

### Versions

N/A

**Comment by Chillee** [comment]
![image](https://github.com/pytorch/pytorch/assets/6355099/eaeea866-7e7a-4664-8e96-a71788d7a74f)

With coordesc tuning the non-welford layernorm strictly dominates the welford layernorm in all cases here.

**Comment by Chillee** [comment]
In fp32 things seem a lot more lenient overall, but welford still seems worse than non-welford layernorm

![image](https://github.com/pytorch/pytorch/assets/6355099/111fb166-9b6d-4395-8244-1088266b6232)

**Comment by peterbell10** [comment]
After investigating further, the main issue is that welford requires more computation per-element than the two pass reduction. For float32 it only uses ~20% of available compute and the kernel is still bandwidth limited, however for half precision we have half the bandwidth requirement and twice as many elements per thread so compute becomes a bottleneck.

I also see a comment in nsight-compute about "Uncoalesced shared accesses" which I think corresponds to an unnecessary layout conversion that I'm seeing in the welford kernel's triton gpu IR. I'll try benchmarking some custom ttgir to see if that could be significant.

I will note that I didn't see quite the same zig-zag performance graphs as @Chillee. My results were roughly monotonically increasing with size. I also think it's a bit suspicious that all the slowest kernels took 4000 ms in the graph above.

**Comment by Chillee** [comment]
I agree my perf results look very suspicious 🤔 But they seem to repro for me.

Can you post the performance you're seeing?

**Comment by peterbell10** [comment]
Yup I think my results are about the same. I'm using the latest triton pin as well
![welford_compare](https://github.com/pytorch/pytorch/assets/13238737/b754e707-ac8f-4b30-ad92-2421f995d14c)

**Comment by manman-ren** [comment]
Some simple questions :]
How do we turn off welford reductions to get the plots? I assume the plots are comparing welford (which is the default) against two-pass (which is turning off welford reductions). And we are seeing this perf issue with bf16 more than with float32.
How to enable coordesc tuning?
Thanks!

**Comment by Chillee** [comment]
@manman-ren https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L5015

setting this function to always return True makes us use non welford reduction.

To enable coordesc tuning do `TORCHINDUCTOR_COORDINATE_DESCENT_TUNING=1`

**Comment by Chillee** [comment]
@manman-ren In particular, some study on when we should use welford reduction vs non welford-reduction (and why) would be very useful to us.

LayerNorm is a very critical operator in deep learning, so our performance on this is quite important :)

**Comment by manman-ren** [comment]
Thanks for the info! I will take a look.

**Comment by manman-ren** [comment]
My numbers are different, but I am on the latest Triton open source (commit 74f1d52f356d2530f797ca10729bbc0b4fb94a4d)
```
size two-pass welford
8192 4278.406143188477 3998.250961303711 -6.54812%
16384 8424.386978149414 8420.171737670898 -0.0500362%
24576 14846.714973449707 12962.637901306152 -12.6902%
32768 18668.025970458984 18799.78370666504 0.705794%
40960 25285.003662109375 26821.632385253906 6.07723%
49152 33661.50665283203 34251.583099365234 1.75297%
57344 41655.58624267578 39714.53094482422 -4.65977%
65536 60410.75134277344 45762.62664794922 -24.2475%
73728 67600.73852539062 51712.41760253906 -23.5032%
81920 75333.56475830078 57604.129791259766 -23.5346%
90112 83022.88055419922 64482.818603515625 -22.3313%
```
I am using
```
for D in range(8192, 8192*11+1, 8192): #256
    inp = torch.randn(2048*128, D, dtype=torch.bfloat16, device='cuda')
    mod = nn.LayerNorm([D], dtype=torch.bfloat16, device='cuda')

    comp_mod = torch.compile(mod, dynamic=False)
    print(D)
    #bench(lambda: mod(inp))
    bench(lambda: comp_mod(inp))
```
changing use_two_step_variance to return True for "two-pass" or return False for "welford". I assume we only need bench for comp_mod?

Checking ttgir for welford (D is 4096)
```
    %6:4 = scf.for %arg8 = %c0_i32 to %c4096_i32 step %c2048_i32 iter_args(%arg9 = %cst_4, %arg10 = %cst_4, %arg11 = %cst_3, %arg12 = %cst_4) -> (tensor<1x2048xf32, #blocked>, tensor<1x2048xf32, #blocked>, tensor<1x2048xf32, #blocked1>, tensor<1x2048xf32, #blocked>)  : i32 {
      %30 = tt.splat %arg8 : i32 -> tensor<1x2048xi32, #blocked> loc(#loc8)
      %31 = arith.addi %30, %2 : tensor<1x2048xi32, #blocked> loc(#loc8)
      %32 = arith.addi %31, %4 : tensor<1x2048xi32, #blocked> loc(#loc5)
      %33 = tt.addptr %5, %32 : tensor<1x2048x!tt.ptr<bf16, 1>, #blocked>, tensor<1x2048xi32, #blocked> loc(#loc6)
      %34 = tt.load %33 {cache = 1 : i32, evict = 3 : i32, isVolatile = false} : tensor<1x2048xbf16, #blocked> loc(#loc9)
      %35 = arith.extf %34 : tensor<1x2048xbf16, #blocked> to tensor<1x2048xf32, #blocked> loc(#loc10)
      %36 = arith.cmpi eq, %arg8, %c0_i32 : i32 loc(#loc11)
      %37:4 = scf.if %36 -> (tensor<1x2048xf32, #blocked1>, tensor<1x2048xf32, #blocked>, tensor<1x2048xf32, #blocked>, tensor<1x2048xf32, #blocked>) {
        scf.yield %cst_1, %35, %cst_4, %cst_0 : tensor<1x2048xf32, #blocked1>, tensor<1x2048xf32, #blocked>, tensor<1x2048xf32, #blocked>, tensor<1x2048xf32, #blocked> loc(#loc67)
      } else {
        %38 = arith.subf %35, %arg9 : tensor<1x2048xf32, #blocked> loc(#loc68)
        %39 = arith.addf %arg12, %cst_0 : tensor<1x2048xf32, #blocked> loc(#loc69)
        %40 = arith.addf %arg11, %cst_1 : tensor<1x2048xf32, #blocked1> loc(#loc69)
        %41 = arith.divf %38, %39 : tensor<1x2048xf32, #blocked> loc(#loc70)
        %42 = arith.addf %arg9, %41 : tensor<1x2048xf32, #blocked> loc(#loc71)
        %43 = arith.subf %35, %42 : tensor<1x2048xf32, #blocked> loc(#loc72)
        %44 = arith.mulf %38, %43 : tensor<1x2048xf32, #blocked> loc(#loc73)
        %45 = arith.addf %arg10, %44 : tensor<1x2048xf32, #blocked> loc(#loc74)
        scf.yield %40, %42, %45, %39 : tensor<1x2048xf32, #blocked1>, tensor<1x2048xf32, #blocked>, tensor<1x2048xf32, #blocked>, tensor<1x2048xf32, #blocked> loc(#loc74)
      } loc(#loc66)
      scf.yield %37#1, %37#2, %37#0, %37#3 : tensor<1x2048xf32, #blocked>, tensor<1x2048xf32, #blocked>, tensor<1x2048xf32, #blocked1>, tensor<1x2048xf32, #blocked> loc(#loc22)
    } loc(#loc7)
    %7 = triton_gpu.convert_layout %6#2 : tensor<1x2048xf32, #blocked1> -> tensor<1x2048xf32, #blocked> loc(#loc7)
```
The convert_layout at the bottom seems to be redundant. But by removing the convert_layout, it seems to have difference perf impact for different value of D, varying from +4% to -4%.

**Comment by Chillee** [comment]
@manman-ren interesting - you're seeing much better welford results than I am. You can also try Peter's PR here: https://github.com/pytorch/pytorch/pull/120330

**Comment by manman-ren** [comment]
Created https://github.com/openai/triton/pull/3510 to handle the extra convert_layout, the perf results are in the PR. The perf results are for welford only (i.e forcing two_pass to be false). The numbers are measured without Peter's PR.
One thing I noticed is that when the kernel's perf is improved with PR3510, the new_config will not be added (https://github.com/pytorch/pytorch/blob/main/torch/_inductor/triton_heuristics.py#L292), causing the perf to be worse compared to the base (without PR3510).
i.e without PR3510, it has two tuning configs 1x2048 and 1x1024, with PR3510, since it runs faster, it only has one tuning config 1x2048
The numbers in PR3510 are collected by forcing the new_config always.

**Comment by manman-ren** [comment]
I wonder if anyone has tuned the tuning configs for welfordreduce for different Ds.

**Comment by peterbell10** [comment]
> interesting - you're seeing much better welford results than I am. 

Does your triton version include openai/triton#3217? Without this there were extraneous layout conversions inside reduction loop which were a big hit to perf.

**Comment by manman-ren** [comment]
Yes it should. My hash is from Mar 26 17:31:55 2024.
Interesting that you worked on the if op! I have https://github.com/openai/triton/pull/3510 that tries to remove a redundant convertop related to results of for op and maybe you have some feedback on it.

**Comment by Chillee** [comment]
> I wonder if anyone has tuned the tuning configs for welfordreduce for different Ds.

If you turn on coordinate descent tuning my experience is that it generally finds pretty good configs.

**Comment by peterbell10** [comment]
I believe you mean the [**naïve algorithm**](https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Na%C3%AFve_algorithm), which as wikipedia states is terribly imprecise and usually shouldn't be used in practice. If it has enough precision for your use case then that's great, but we can't use it as the general variance implementation.

**Comment by sharpobject** [comment]
I'll be landing some improvements to LayerNorm for cpu which greatly reduce the cost of welford's algorithm, including by removing all of the divides except for the single divide at the end of the reduction. So I hope that in the near future I can get these into GPU as well, and there won't be any need to compromise by using a less numerically stable solution.
